File: platform/simulation/src/generation/builders/xml_helpers.py
"""Common XML element construction utilities for SDF builders."""

from __future__ import annotations

from xml.etree import ElementTree as ET


def add_box_visual(
    link: ET.Element,
    width: float,
    depth: float,
    height: float,
    color: list[float] | tuple[float, ...],
    visual_name: str = "visual",
) -> None:
    """Add a box visual element to a link.

    Args:
        link: The <link> ET element to add the visual to.
        width: Box width (X dimension).
        depth: Box depth (Y dimension).
        height: Box height (Z dimension).
        color: RGB color list/tuple normalized to [0, 1].
        visual_name: Name attribute for the visual element.
    """
    vis = ET.SubElement(link, "visual", name=visual_name)
    geom = ET.SubElement(ET.SubElement(vis, "geometry"), "box")
    ET.SubElement(geom, "size").text = f"{width} {depth} {height}"
    mat = ET.SubElement(vis, "material")
    color_str = f"{color[0]} {color[1]} {color[2]} 1"
    ET.SubElement(mat, "ambient").text = color_str
    ET.SubElement(mat, "diffuse").text = color_str


def add_box_collision(
    link: ET.Element,
    width: float,
    depth: float,
    height: float,
    collision_name: str = "collision",
) -> None:
    """Add a box collision element to a link.

    Args:
        link: The <link> ET element to add the collision to.
        width: Box width (X dimension).
        depth: Box depth (Y dimension).
        height: Box height (Z dimension).
        collision_name: Name attribute for the collision element.
    """
    col = ET.SubElement(link, "collision", name=collision_name)
    geom = ET.SubElement(ET.SubElement(col, "geometry"), "box")
    ET.SubElement(geom, "size").text = f"{width} {depth} {height}"


def build_wall_model(
    name: str,
    center_x: float,
    center_y: float,
    visual_x: float,
    visual_y: float,
    collision_x: float,
    collision_y: float,
    z_position: float = 0.0,
) -> ET.Element:
    """Construct a static wall model element.

    Args:
        name: Model name.
        center_x: X position of wall center.
        center_y: Y position of wall center.
        visual_x: Visual geometry X dimension.
        visual_y: Visual geometry Y dimension.
        collision_x: Collision geometry X dimension.
        collision_y: Collision geometry Y dimension.
        z_position: Z position of wall (default: 0.0).

    Returns:
        A constructed <model> ET element.
    """
    from shared.config.constants import WallSpecs

    model = ET.Element("model", name=name)
    ET.SubElement(model, "static").text = "true"
    ET.SubElement(model, "pose").text = f"{center_x} {center_y} {z_position} 0 0 0"

    link = ET.SubElement(model, "link", name="link")

    # Visual (thin appearance)
    add_box_visual(
        link,
        visual_x,
        visual_y,
        WallSpecs.HEIGHT,
        WallSpecs.COLOR,
        visual_name="visual",
    )

    # Collision (thicker for safety)
    add_box_collision(
        link,
        collision_x,
        collision_y,
        WallSpecs.HEIGHT,
        collision_name="collision",
    )

    return model

File: platform/gazebo/generator/internal/sdf/xml_helpers.go
package sdf

import (
	"fmt"
	"strconv"

	"voldemorbot/gazebo/generator/internal/simconfig"
)

// Axis direction constants for joint and geometry normal definitions.
const (
	axisX = "1 0 0"
	axisY = "0 1 0"
	axisZ = "0 0 1"
)

// Joint limit sentinels for continuous revolute joints (±1e16 = no limit in Gazebo SDF).
const (
	jointLimitUnbounded    = "1e16"
	jointLimitUnboundedNeg = "-1e16"
)

// Zero-value string constants for SDF elements.
const (
	floatZero = "0.0"
	intZero   = "0"
)

// ff formats a float64 as a decimal string without trailing zeros.
func ff(f float64) string {
	return strconv.FormatFloat(f, 'f', -1, 64)
}

// fi formats an integer as a decimal string.
func fi(i int) string {
	return strconv.Itoa(i)
}

// pose6 formats a 6-DOF pose string "x y z roll pitch yaw".
func pose6(x, y, z, roll, pitch, yaw float64) string {
	return fmt.Sprintf("%s %s %s %s %s %s", ff(x), ff(y), ff(z), ff(roll), ff(pitch), ff(yaw))
}

// poseXY formats a pose at (x, y, 0) with zero rotation.
func poseXY(x, y float64) string {
	return fmt.Sprintf("%s %s 0 0 0 0", ff(x), ff(y))
}

// vec3 formats three floats as "x y z".
func vec3(x, y, z float64) string {
	return fmt.Sprintf("%s %s %s", ff(x), ff(y), ff(z))
}

// rgba formats an RGB triple plus alpha 1 as "r g b 1".
func rgba(c simconfig.RGB) string {
	return fmt.Sprintf("%s %s %s 1", ff(c[0]), ff(c[1]), ff(c[2]))
}

// AddBoxVisual appends a box <visual> element to a link node.
func AddBoxVisual(link *Node, width, depth, height float64, color simconfig.RGB, visualName string) {
	if visualName == "" {
		visualName = "visual"
	}
	vis := link.Sub("visual", "name", visualName)
	vis.Sub("geometry").Sub("box").SubT("size", vec3(width, depth, height))
	mat := vis.Sub("material")
	c := rgba(color)
	mat.SubT("ambient", c)
	mat.SubT("diffuse", c)
}

// AddBoxCollision appends a box <collision> element to a link node.
func AddBoxCollision(link *Node, width, depth, height float64, collisionName string) {
	if collisionName == "" {
		collisionName = "collision"
	}
	col := link.Sub("collision", "name", collisionName)
	col.Sub("geometry").Sub("box").SubT("size", vec3(width, depth, height))
}

// BuildWallModel constructs a static wall <model> element from center coords and dimensions.
func BuildWallModel(name string, cx, cy, visualX, visualY, collisionX, collisionY float64) *Node {
	model := New("model", "name", name)
	model.SubT("static", "true")
	model.SubT("pose", pose6(cx, cy, 0, 0, 0, 0))
	link := model.Sub("link", "name", "link")
	AddBoxVisual(link, visualX, visualY, simconfig.WallHeight, simconfig.WallColor, "visual")
	AddBoxCollision(link, collisionX, collisionY, simconfig.WallHeight, "collision")
	return model
}
